src/compiler/checker.ts

```diff
@@ -11432,6 +11436,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 : declaredType ? declaredType
                 : getAllSymbolFlags(targetSymbol) & SymbolFlags.Value ? getTypeOfSymbol(targetSymbol)
                 : errorType;
+
+            if (!popTypeResolution()) {
+                if (exportSymbol) {
+                    reportCircularityError(exportSymbol);
+                }
+                else {
+                    reportCircularityError(symbol);
+                }

```

Comment:
This is shorter, and I *think* it's still correct:
```suggestion
                reportCircularityError(exportSymbol ?? symbol);
```